### PR TITLE
[icp] fpu and asm cleanup for linux

### PR DESCRIPTION
**Pull request by ironMann**
### Description
Properly annotate functions and data section so that objtool does not complain
when CONFIG_STACK_VALIDATION and CONFIG_FRAME_POINTER are enabled.
Pass KERNELCPPFLAGS to assembler.
Use kfpu_begin()/kfpu_end() to protect SIMD regions in Linux kernel.

Issue #5041

### How Has This Been Tested?
This has not been tested!  

### Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [x] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Checklist:
- [x] My code follows the ZFS on Linux code style requirements.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] Change has been approved by a ZFS on Linux member.

### CODE REVIEW

File: config/kernel-objtool.m4 [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+dnl # 4.6 API for compile-time stack validation
+dnl #
+AC_DEFUN([ZFS_AC_KERNEL_OBJTOOL], [
+	AC_MSG_CHECKING([for compile-time stack validation (objtool)])
+	ZFS_LINUX_TRY_COMPILE([
+		#undef __ASSEMBLY__
+		#include <asm/frame.h>
+	],[
+		#if !defined(FRAME_BEGIN)
+		CTASSERT(1);
```
Comment (behlendorf) [inline config/kernel-objtool.m4]: nit: using `#error` would be consistent with the rest of the m4 macros.

**Comment by ironMann** [comment]
@behlendorf I dont get warnings for avx512 with Fedora 4.8 kernel. `objtool` appeared in 4.6, and was made aware of x86 EVEX prefix instructions (used for AVX512) in linux 4.8. 
So I suspect those are false positives.

**Comment by behlendorf** [comment]
The posted warnings are from a CentOS 7 `3.10.0-514.6.1.el7.x86_64` kernel.  It appears RedHat has pulled in objtool but not all of the upstream fixes.  I believe torvalds/linux@8d94c2f9198301bd78feb1d9ad6c08a8ee7b5d1c is the upstream fix you're describing.  So it does look like these are false positives and we'll want RedHat to pull in the fix.

**Comment by tcaputi** [comment]
We added the patch to #5769 and all of the automated tests passed with no issues. I'm not an assembly expert, but I can see that algorithmically nothing has changed. If someone else can verify the register managing changes that @ironMann has made, I'm fine with it.

**Comment by behlendorf** [comment]
@tcaputi thanks for the quick verification.

**Pull request merged**